src/libPOW/pow.cpp

```diff
@@ -238,8 +242,250 @@ ethash_mining_result_t POW::MineFullGPU(uint64_t blockNum,
+      DataConversion::IntegerToBytes<uint32_t, sizeof(uint32_t)>(
+          POW_WINDOW_IN_SECONDS);
+  tmp.insert(tmp.end(), powTimeBytes.begin(), powTimeBytes.end());
+
+  if (tmp.size() != (PUB_KEY_SIZE + BLOCK_HASH_SIZE + sizeof(uint64_t) +
+                     BLOCK_HASH_SIZE + sizeof(uint32_t))) {
+    LOG_GENERAL(WARNING, "Size of the buffer "
+                             << tmp.size()
+                             << " to generate signature is not correct.");
+    return false;
+  }
+
+  Signature signature;
+  if (!Schnorr::GetInstance().Sign(tmp, pairOfKey.first, pairOfKey.second,
+                                   signature)) {
+    LOG_GENERAL(WARNING, "Failed to sign zil_requestWork json value.");
+    return false;
+  }
+  jsonValue[5] = "0x" + DataConversion::SerializableToHexStr(signature);
+
+  LOG_GENERAL(INFO, "Json value send out: " << jsonValue);
+
+  try {
+    jsonrpc::Client client(*m_httpClient);
+    Json::Value ret = client.CallMethod("zil_requestWork", jsonValue);
+    LOG_GENERAL(INFO, "zil_requestWork return: " << ret);
+    return ret.asBool();
+  } catch (std::exception& e) {
+    LOG_GENERAL(WARNING,
+                "Exception captured in jsonrpc api zil_requestWork, exception: "
+                    << e.what());
+    return false;
+  }
+}
+
+bool POW::CheckMiningResult(const PairOfKey& pairOfKey, uint64_t blockNum,
+                            ethash_hash256 const& headerHash,
+                            ethash_hash256 const& boundary,
+                            ethash_mining_result_t& miningResult) {
+  Json::Value jsonValue;
+
+  bytes tmp;
+  bytes pubKeyData;
+
+  pairOfKey.second.Serialize(pubKeyData, 0);
+  jsonValue[0] = "0x" + BytesToHexString(pubKeyData.data(), pubKeyData.size());
+  tmp.insert(tmp.end(), pubKeyData.begin(), pubKeyData.end());
+
+  jsonValue[1] = "0x" + BlockhashToHexString(headerHash);
+  tmp.insert(tmp.end(), headerHash.bytes,
+             headerHash.bytes + sizeof(ethash_hash256));
+
+  jsonValue[2] = "0x" + BlockhashToHexString(boundary);
+  tmp.insert(tmp.end(), boundary.bytes,
+             boundary.bytes + sizeof(ethash_hash256));
+
+  Signature signature;
+  if (!Schnorr::GetInstance().Sign(tmp, pairOfKey.first, pairOfKey.second,
+                                   signature)) {
+    LOG_GENERAL(WARNING, "Failed to sign zil_checkWorkStatus json value.");
+    return false;
+  }
+  jsonValue[3] = "0x" + DataConversion::SerializableToHexStr(signature);
+
+  LOG_GENERAL(INFO, "Json value send out: " << jsonValue);
+
+  while (m_shouldMine) {
+    std::this_thread::sleep_for(
+        std::chrono::seconds(CHECK_MINING_RESULT_INTERVAL));
+
+    try {
+      jsonrpc::Client client(*m_httpClient);
+      Json::Value ret = client.CallMethod("zil_checkWorkStatus", jsonValue);
+      LOG_GENERAL(INFO, "zil_checkWorkStatus return: " << ret);
+      bool workDone = ret[0].asBool();
+      if (!workDone) {
+        continue;
+      }
+
+      if (ret.size() < 4) {
```

Comment:
can this four be some constant ? like NUM_JSON_ELEMENTS

Reply:
No need, this is defined in the communication protocal, https://docs.google.com/document/d/13KwqovblQ4gsZnXMsP6Lw7Yl3F5uzD6VqBheHgXB1wQ/edit

Reply:
yeah, but i feel it is better as a constant, maybe one defined in this file. no need to define in constants.xml

Reply:
Ok, i use constant value already